tests/FSharp.Compiler.Service.Tests/TypedTreePickleTests.fs

```diff
@@ -0,0 +1,166 @@
+﻿module FSharp.Compiler.Service.Tests.TypedTreePickleTests
+
+open System.IO
+
+open FSharp.Compiler
+open FSharp.Compiler.AbstractIL.IL
+open FSharp.Compiler.AbstractIL.ILBinaryReader
+open FSharp.Compiler.CodeAnalysis
+open FSharp.Compiler.CompilerConfig
+open FSharp.Compiler.CompilerImports
+open FSharp.Compiler.DependencyManager
+open FSharp.Compiler.Text.Range
+open FSharp.Compiler.Driver
+open FSharp.Compiler.TypedTree
+open FSharp.Compiler.TypedTreePickle
+open FSharp.Compiler.Syntax
+
+open Internal.Utilities.Library
+
+open Xunit
+
+type TypedTreePickleTests() =
+
+    /// random auxiliary stuff
+
+    let builder = TcConfigBuilder.CreateNew(
+        SimulatedMSBuildReferenceResolver.getResolver(),
+        Directory.GetCurrentDirectory(),
+        ReduceMemoryFlag.No,
+        "",
+        false,
+        false,
+        CopyFSharpCoreFlag.No,
+        (fun _ -> None),
+        None,
+        range0,
+        compressMetadata = false)
+
+    let tcConfig = TcConfig.Create(builder, false)
+
+    let sysRes, otherRes, knownUnresolved =
+        TcAssemblyResolutions.SplitNonFoundationalResolutions(tcConfig)
+
+    let foundationalTcConfigP = TcConfigProvider.Constant tcConfig
+    let tcGlobals, frameworkTcImports = 
+        TcImports.BuildFrameworkTcImports(
+            foundationalTcConfigP,
+            sysRes,
+            otherRes) 
+        |> Async.RunImmediate
+
+    let tcImports =
+        TcImports.BuildNonFrameworkTcImports(
+            foundationalTcConfigP,
+            frameworkTcImports, 
+            otherRes, 
+            knownUnresolved, 
+            new DependencyProvider())
+        |> Async.RunImmediate
+
+    let tcEnv0, openDecls0 = ParseAndCheckInputs.GetInitialTcEnv(
+        "test",
+        rangeStartup,
+        tcConfig, 
+        tcImports,
+        tcGlobals)
+
+    /// helper functions
+
+    let prepareData sourceFiles =
+        let inputs =
+            ParseAndCheckInputs.ParseInputFiles(
+                tcConfig, 
+                Lexhelp.LexResourceManager(), 
+                sourceFiles, 
+                DiagnosticsLogger.AssertFalseDiagnosticsLogger, 
+                false) |> List.map fst
+
+        let tcState, topAttribs, implFiles, _ = 
+            TypeCheck(
+                CompilationThreadToken(),
+                tcConfig,
+                tcImports,
+                tcGlobals,
+                DiagnosticsLogger.AssertFalseDiagnosticsLogger,
+                "test",
+                tcEnv0,
+                openDecls0,
+                inputs,
+                DiagnosticsLogger.QuitProcessExiter,
+                "")
+
+        let tcInfo : PickledTcInfo = {
+            MainMethodAttrs = topAttribs.mainMethodAttrs
+            NetModuleAttrs = topAttribs.netModuleAttrs
+            AssemblyAttrs = topAttribs.assemblyAttrs
+            DeclaredImpls = implFiles
+        }
+
+        tcState.Ccu, tcInfo
+
+
+    [<Fact>]
+    let PickleAndUnpickleTcData() =
```

Comment:
More generally, how to test this?

Comparing the original and restored TC information is hard currently since many types don't have equality allowed. Should we change this just for the purpose of testing? Doesn't feel right.

What else - create baselines for resources (bytes)? 

Rely on the higher-level testing?

Hope & pray? (like we do with other pickling & unpickling stuff...)

Reply:
E2e testing of an integrated scenario.

Which can then compare e.g. IL baseline and separately compare debug info data, their equality is easier tested. Also, those should not be deviating anyway.

The same way pickling and unpickling is tested - it is not tested by hope & pray, it is tested by integration tests using referenced pickled version of F# code, such as the code in FSharp.Core and its signature and optimization data. (e.g. integration tests proving that Fsharp.Core calls are being correctly inlined thanks to optimization data)